frame/header-mmr/src/lib.rs

```diff
@@ -172,19 +179,18 @@ impl<T: Trait> MMRStore<T::Hash> for ModuleMMRStore<T> {
 impl<T: Trait> Module<T> {
 	impl_rpc! {
 		pub fn gen_proof_rpc(
-			block_number: T::BlockNumber,
-			mmr_block_number: T::BlockNumber,
+			block_number_of_member_leaf: u64,
+			block_number_of_last_leaf: u64,
 		) -> RuntimeDispatchInfo<T::Hash> {
-			if block_number < mmr_block_number {
-				let pos = Self::position_of(block_number);
-				let mmr_header_pos = Self::position_of(mmr_block_number);
-
+			if block_number_of_member_leaf < block_number_of_last_leaf {
 				let store = <ModuleMMRStore<T>>::default();
-				let mmr = <MMR<_, MMRMerge<T>, _>>::new(mmr_header_pos, store);
+				let mmr_size = block_number_to_mmr_size(block_number_of_last_leaf);
+				let mmr = <MMR<_, MMRMerge<T>, _>>::new(mmr_size, store);
+				let pos = block_number_to_pos(block_number_of_member_leaf);
```

Comment:
```block_number_to_pos``` aka ```leaf_index_to_pos ``` from ckb mmr lib is not optimized, it can actually calculated by 2N - k - H - 1 or to_mmr_size(block) - H - 1, H is suffix zero counts(from last 1, exclusive)

ckb's version is ok to go for now.

Reply:
I've tried this algorithm yesterday morning:

```rust
pub fn pos(block_count: u64) -> u64 {
    // if suffix zero counts from last 1
    Self::mmr_size(block_count) - (((block_count + 1) >> 1).trailing_zeros() as u64) + 1
    // if suffix zero counts from last 0
    // Self::mmr_size(block_count) - ((block_count + 1).trailing_zeros() as u64) + 1
}
```

I'm not sure if my implementation is wrong, but it didn't work.

Reply:
The `Self::mmr_size` in that implementation is as same as `block_number_to_mmr_size` here, this algorithm works fine, tested with index (0..10)

Reply:
> I've tried this algorithm yesterday morning:
> 
> ```rust
> pub fn pos(block_count: u64) -> u64 {
>     // if suffix zero counts from last 1
>     Self::mmr_size(block_count) - (((block_count + 1) >> 1).trailing_zeros() as u64) + 1
>     // if suffix zero counts from last 0
>     // Self::mmr_size(block_count) - ((block_count + 1).trailing_zeros() as u64) + 1
> }
> ```
> 
> I'm not sure if my implementation is wrong, but it didn't work.

https://github.com/darwinia-network/darwinia-common/blob/aefdad9944e4a5c7611db3ba5892aa373ad1033c/frame/header-mmr/src/tests.rs#L148